### DISCUSSION TITLE
The tree printed by `lexy::trace` shows successful parsing, but still reports errors.

### DISCUSSION THREAD

**Discussion by Life4gal** [foonathan/lexy#119] [Q&A] [answered]
I am using `lexy` to write a simple [`ini`](https://github.com/Life4gal/ini) file parsing library, it works fine, but there is a problem.

![image](https://user-images.githubusercontent.com/52756109/215967833-9f472a02-bcb8-4e6a-a031-b9339beed8f0.png)
My intention is to support optional `value` (`key` is necessary) and optional `inline comment`.

In order not to have the comment starters (`;` and `#`) be considered part of the `value`, my `value` looks like this:
![image](https://user-images.githubusercontent.com/52756109/215968527-54fe1cb7-eba9-488e-9661-f58b98ae04d7.png)

This allows adding `inline comment` after a key-value pair that has no `value`.

But this presents a problem, as the title suggests.
![image](https://user-images.githubusercontent.com/52756109/215968918-09c550a8-cefe-4fca-be12-91f07f18beec.png)
![image](https://user-images.githubusercontent.com/52756109/215968962-96df761b-bd47-4320-a31f-fc6a85a4a0cd.png)
![image](https://user-images.githubusercontent.com/52756109/215969054-d8636be3-93d6-4675-97ac-a77e4ab77281.png)

And it does fail, because the `key4` line does not repeatedly write `# foo bar baz here` as the other lines do.

If you need further information, please contact me. :)

**Comment by foonathan** [comment]
That seems like a bug. Can you share a playground example that reproduces it? https://lexy.foonathan.net/playground/

**Reply by Life4gal** [reply]
Because the tests use both `extractor` and `flusher`, the test case needs to use nearly the entire project.

Regardless of the problems with [linux&clang due to the constructor](https://github.com/Life4gal/ini/actions/runs/4061564842/jobs/6991709002#step:7:24) and [win&msvc due to cmake passing -std=c++20](https://github.com/Life4gal/ini/actions/runs/4061564842/jobs/6991709551#step:8:68), the nested namespace is not recognized correctly. 

It is strange that win&clang have different results due to `debug|release`.
And in the release version, [the parsing does fail.](https://github.com/Life4gal/ini/actions/runs/4061564842/jobs/6991709480#step:9:946)

If you want, you can try to test [my project](https://github.com/Life4gal/ini) by compiling it directly. :)

**Reply by foonathan** [reply]
The playground doesn't need value callbacks, you should be able to copy paste your grammar and remove all `::value` members, and it should compile there. It would make a fix a lot quicker ;)

**Reply by Life4gal** [reply]
https://lexy.foonathan.net/playground/?id=nvb5Kcajq&mode=tree

As expected, both `parse tree` and `trace` are correct, after all it only parses once. 😄 

-------------------------------------------------------------------
After debugging, I am not sure where the problem, it seems to be `StackFunction` problem, I tested the configuration of `win&clang`, found that `Debug/RelWithDebInfo/MinSizeRel` are no problem, but `Release` will cause a crash.

**Reply by Life4gal** [reply]
It seems to be a problem with `parse`.

[See this](https://github.com/Life4gal/ini/blob/733fab4c8d90f190f8595cead116109bd12b4496/src/flusher.cpp#L529)

I did the `trace` before the `parse`. The `trace` reports an error, and after that, the `parse` causes the crash.

**Reply by Life4gal** [reply]
The crash was found, I didn't set the default value for comment::indication, which caused the `Release` version not to be guaranteed to be `INVALID` and caused it to execute in the [`UNREACHABLE` branch](https://github.com/Life4gal/ini/blob/737d1176c025bd28fe2d6ea95f57aad36b297095/include/ini/internal/common.hpp#L507).

Now I can't reproduce the previous problem, so I hope this is the solution.

----------------------------------------------------------
I am very skeptical about this solution, and I still can't understand why `trace` is reporting errors. I don't even think this crash is related to the `parse` problem.

As I have shown in the screenshot above, there was a problem with one of the lines in the middle of the parsing (the key4 line) and the line was skipped and the parsing continued. The result is that except for the key4 line, all other lines are as expected. This obviously can not be caused by a program crash.

**Reply by foonathan** [reply]
So to clarify, in your linked playground example, the parse tree and trace is correct, but when you use the parse action itself, you get an error message?

**Reply by Life4gal** [reply]
Yes, it is very strange.

**Reply by foonathan** [reply]
I've cloned and compiled your project, all test pass for me.

**Reply by Life4gal** [reply]
The test passes, but the content of the file is not as expected.

![image](https://user-images.githubusercontent.com/52756109/217492016-55d78fa6-246e-4581-8034-72ccb61774df.png)
![image](https://user-images.githubusercontent.com/52756109/217492082-b252e2b7-9f6a-4c1f-9ec8-4b3c88e622a9.png)

commit: 733fab4c8d90f190f8595cead116109bd12b4496

**Reply by Life4gal** [reply]
This issue is also present in the latest commit.

![image](https://user-images.githubusercontent.com/52756109/217492856-f76f55af-a89a-4655-ab14-7581cd5f65e3.png)

**Reply by Life4gal** [reply]
So the problem comes back to the `variable_declaration` I mentioned above. :)

**Reply by foonathan** [reply]
Please give me a C++ source files that contains

* The input as a string literal.
* A call to `lexy::trace` that works as expected.
* A call to `lexy::parse` that fails.

I can't really help you otherwise.

**Reply by Life4gal** [reply]
Oh! I noticed that the `trace` after

![image](https://user-images.githubusercontent.com/52756109/217495306-80654730-79a7-45f0-b08d-929a977f56bf.png)

`key4` did fail to parse. But then there is another parsing that is successful, I got them mixed up.

**Reply by Life4gal** [reply]
> Please give me a C++ source files that contains
> 
> * The input as a string literal.
> * A call to `lexy::trace` that works as expected.
> * A call to `lexy::parse` that fails.
> 
> I can't really help you otherwise.

unit_test/src/test_flusher.cpp

**Reply by foonathan** [reply]
That file contains a bunch of other code. It also isn't at all clear how it generates the input and where the actual call happens. I want something like:

```cpp
#include <ini/foo.hpp>

int main()
{
     auto input = "...";
     
     lexy::trace(...);
     lexy::parse(...);
}
```

**Reply by Life4gal** [reply]
I will try, but it may be a bit complicated, I now suspect because there are some unexpected differences between my `flusher` and `extractor`'s grammar.

**Reply by Life4gal** [reply, marked as answer]
After testing, I'm sure it's the inconsistent grammar of extractor and flusher that's causing the problem. :)

**Reply by foonathan** [reply]
Let me know if you need my help with anything.
